R/utils.R

```diff
@@ -488,7 +488,13 @@ shinyCallingHandlers <- function(expr) {
         return()
 
       handle <- getOption('shiny.error')
-      if (is.function(handle)) handle()
+      if (is.function(handle)) {
+        if ("error" %in% names(formals(handle))) {
+          handle(error = e)
```

Comment:
I think there's less chance of confusion if, instead of looking for an argument named `error`, this simply checked that the function takes at least one argument, and then invoked `handler(e)`.

This is because it's very common to write handlers like this:

```R
tryCatch(
  ...,
  error = function(e) { ... }
)
```

So it's likely that someone will try writing a handler like this:

```R
options(shiny.error = function(e) { ... })
```

If the code here were to throw an error due to being invoked as `handler()` (without being provided a value for `e`), that would be quite surprising and confusing for the user.

Reply:
Good point about `browser` and `traceback` taking arguments. Is there some way we can make it easier for people to detect and correct this problem if they do provide a `function(e) { ... })`? My first instinct for writing an error handler would be to write it as `function(e)` instead of `function(error)` and I'm sure I'm not alone here -- and if someone does that, it would be best if we could help them onto the right path, otherwise they might just be confused by what happens.

For example, with this app:

```R
options(shiny.error = function(e) {
  print(e)
})

shinyApp(
  ui = fluidPage(),
  server = function(input, output) {
    observe({
      stop("Error here!")
    })
  }
)
```

This is the output:

```
Listening on http://127.0.0.1:5769
Warning: Error in handle: argument "e" is missing, with no default
  54: print
  53: handle [#2]
  52: <Anonymous> [/Users/winston/Projects/shiny/R/utils.R#495]
  51: stop
  50: h [/Users/winston/Projects/shiny/R/conditions.R#209]
  49: .handleSimpleError
  48: stop
  47: observe [#4]
  46: <observer> [/Users/winston/Projects/shiny/R/utils.R#1466]
   3: runApp [/Users/winston/Projects/shiny/R/runapp.R#388]
   2: print.shiny.appobj [/Users/winston/Projects/shiny/R/shinyapp.R#553]
   1: <Anonymous>
```

I think that if a user doesn't already know what the problem is here, they would probably get stuck here, even after reading the docs. It would help if the docs made it painfully clear that the argument _must_ be named `error`, because it would be easy to miss that fact if you're used to being able to write `function(e)`.